### DISCUSSION TITLE
How to track "in-progress" frontend state?

### DISCUSSION THREAD

**Discussion by adamcunnington-mlg** [bigskysoftware/htmx#1544] [Q&A]
By "in-progress" state, I'm talking about a user interacting on some frontend web page and progressively doing stuff; filling out a form, adding rows to a table, deleting rows etc. etc. etc. Eventually they are going to click "Submit" or "Save" but until that point, I'm trying to figure out how one manages such state in an HTMX-app context where I am trying to avoid writing javascript.

In a SPA, frontend manages all state in browser memory via javascript - redux and the like.
In a MPA, backend manages all state. Frontend posts state via html form. This is straightforward though because the page's structure is not dynamic so the HTML form can keep track of "in-progress" state and then post when ready.

With HTMX, as far as I can understand, we're sort of in a half-way house. Technically, it's an MPA but we're refreshing parts of the page via AJAX and thus, my frontend user-input is of a dynamic structure.

As far as I can tell, every time the user does something in the frontend - iterating on their "in-progress" state, the frontend must send this to the backend and the backend must keep track. The backend could store in session cookie, backend in-memory cache or persistent store (likely not this though as the data is not yet "submitted") but aside from other challenges, I am stumbling over the very first hurdle which is how can the backend actually RELATE state from a particular "context". If I open 3 tabs and have 3 concurrent "in-progress" journeys, however I choose the store the "in-progress" state, I need to be able to know which one relates to which. So I need some sort of deterministic key that the frontend can use to only grab it's state. But what could such a deterministic key be? I've not inserted anything into a database yet and some "request id" isn't good enough. I need "tab id"!

The only option I can think of is perhaps using a websocket - I don't know much about this beyond the general way in which it works - but is that the right way to keep track of progressive state?

I feel like my scenario is a really common one but the solution is really unobvious to me. Is the answer simply "well, this is where you actually do need to use javascript and you should do that alongside htmx - perhaps redux".

Help! Thanks!

**Comment by adamcunnington-mlg** [comment]
Or maybe I'm over complicating this and html forms are fine and I just have to be careful when refreshing PARTS of the form so I don't lose any state. All "in-progress" tab state will be in the DOM and I can just select the bits that I want when posting?

**Reply by Telroshan** [reply]
Just to comment on the "tab id" part. I'm personnally simply generating a UUID (on the backend) for every root URL request (that is, a using landing on a page, but not a user making a partial request to update a portion of the view).
When rendering the initial page, I just define that ID in the client's JS as a global variable, and now the client knows what "tab ID" they have, until closing the tab, or loading another page.
I'm using it mostly for SSE (so websockets but with server -> client communication only and not client -> server) and it works very well! This lets me send updates through SSE to other active tabs for a given user, while ignoring the tab from the user that made the request.
I'm not storing this in a DB, just using a map in RAM

As for the in-progress state thing, I would indeed be in favor of 
> just have to be careful when refreshing PARTS of the form so I don't lose any state

Browsers already have nice behaviors for preserving values, though I'm using Chrome and haven't tested that out on other browsers, but if you look at [this basic JSFiddle](https://jsfiddle.net/tqhg0w52/) and run it, type whatever you want in the inputs ,"accidentaly" go to another URL, but if you hit the back button, the values which hadn't been submitted are still there.

Don't know the exact use case you're trying to cover here but I would really suggest not making this too complicated, if you're worried about the user "losing" values that hadn't been submitted, it's either because they left the tab (and the browser would already take care of that), or because your code wiped out their data (in which case you should just make your code avoid doing such a thing!)

Finally, if you really need to add some logic that isn't there by default, I would advocate for custom JS indeed. I think hand-written JS perfectly makes sense here, as it's for user experience purposes. It's a very sane objective of hand-written JS. I think storing "not submitted" states on the backend would be a bit too overkill, unless you want to provide your user the ability to start entering values on a given device (let's say a PC), then finish filling their form on another device (let's say a smartphone). Other than this usecase, I would suggest doing this client side only.

Hope this helps!

**Reply by adamcunnington-mlg** [reply]
@Telroshan this is really helpful - thanks so much!

Just one question, how does your backend generate a UUID per "tab" so to speak? Specifically what values from the request can you use to do that - and how does the client know that it's "theirs" - or is that what the websocket is achieving for you there? The client "tab" receives only its stuff back because of the socket comms?

**Reply by Telroshan** [reply]
I'm talking about "tabs" but it's really more about "pages". For every page reload, or navigation to another URL, the backend generates a new UUID. This isn't an issue in my case though as the "old" page is dead anyway (and was replaced by the new one), thus doesn't need to be informed of anything new that happened elsewhere.
- By "a page", I'm referring to the very initial request your client makes when accessing the page URL. So when you render a template that has the `html`, `head` & `body` tags, you can consider that as a tab. In this case, I generate a UUID and write it in the rendered HTML, with something like
```html
<script type="text/javascript">
    window.tabUuid = "{{ .TabUuid }}"
</script>
```
Where `{{ .TabUuid }}` is a server instruction _(golang syntax in this case)_, that resolves to the UUID your backend generated
- On the other hand, when you render only a portion of HTML, let's say a single div, you know that it's going to be swapped by htmx, thus isn't an initial page load but a partial request that happens during the page's lifetime. In this case, I do not generate any UUID as it's coming from an already existing "tab"

Then, when setting up the SSE connection (works the same for websockets), I simply retrieve that `window.tabUuid` parameter and put that in the SSE setup request. This way:
- The client knows their tab UUID, which is stored in a client-side global variable. Its value is never going to change during the page's lifetime
- The backend knows that this UUID matches this SSE (or websocket) connection, in a map stored in RAM (I usually also use another map, to associate a User ID to an array of all the active SSE/websocket UUID sessions associated to that user, to be able to target "all active tabs from user X")
- From the client, I can now pass that UUID along any request I make with htmx (you can use [`hx-vals`](https://htmx.org/attributes/hx-vals/) or [`hx-include`](https://htmx.org/attributes/hx-include/) from your HTML, or pass it in the values of [`htmx.ajax`](https://htmx.org/api/#ajax) from your JS)
- In the backend's handlers, I can retrieve the UUID the client sent along their request _(just a regular form / URL-encoded value)_, and basically say "ok now send this websocket/SSE event to every tab that this User has opened, except the one with this UUID". For the client that made the request, I can simply send stuff in the HTTP response as usual.
- All other active tabs get the event, but not the one that made the initial request with htmx

Hope that helps!

**Reply by adamcunnington-mlg** [reply]
Thanks - this is again very helpful!

Final Q - why did you use websockets in your case? Is it critical to the workflow you described or for some other reason?

**Reply by Telroshan** [reply]
To give full context, I only needed it in one of my projects. It's [a project management app](https://zorro.management), in which you can work in team but also just make changes in a tab while having other active tabs.
It seemed critical to me to provide a "live update" feature to users, so that what you're viewing is always up to date and never outdated (tasks can be moved, renamed, other information can be updated, and so on).
So it's really for user experience purposes, to provide a "seamless experience", if that makes sense!

**Reply by adamcunnington-mlg** [reply]
@Telroshan thanks again for this - I'm clear now that you used the websockets for something slightly different.

Back to the core of the question - if I want my backend to track "in-progress state" from the frontend and I want to impose the following deliberate constraints:
* does not use cookie because not big enough
* don't want to use backend in-memory cache because a) expensive and b) i'd have to emulate a lot of behaviour you get for free when using browser session state

... then what are my options? I think the only option would be to use browser local storage (or better still, session storage) - which presumably has a mucher higher limit than 4KB. I don't think HTMX allows me to interact with localStorage directly - but I could do this via javascript or hyperscript?

Again, I can't help but feel that my use case is not that unorthodox but super unclear how to do this in htmx context. I guess in a frontend framework, the framework would do this stuff for you (e.g. redux via browser local storage).

**Reply by Telroshan** [reply]
Just to be sure there,
> if I want my backend to track "in-progress state"

But what you're describing with your constraints & goals, is having instead your frontend track the in-progress state and not the backend at all, or did I miss something?

Could you by any chance provide a step-by-step usecase defining what you're trying to achieve (as the user, the functional process) ? I'm afraid my mid-tier English is getting in the way here and I can't fully grasp what you're trying to do exactly

**Reply by adamcunnington-mlg** [reply]
Sure, thanks for the willingness!

So, imagine I have a pagined table in my frontend - the contents are provided by backend, via HTMX request/response.
I want to allow users to edit details in this table as well as do typical table control stuff - click on header to order etc. Each interaction with the table will send request to backend. Backend will merge the "in-progress" state with actual data and return the results (entire HTML table) accordingly.
Imagine a user has edited a row (which is currently in the DOM) but now clicks on a header that would cause this row to be ordered in a latter page. When the backend returns the updated table (page 1), that edit won't be in the DOM and it will be "lost". 
Therefore, the backend must keep track of this in-progress state. The backend needs a persistent, fast "cache" that follows session semantics (including ability to correlate information to a given "tab id" - you've solved this by backend persisting some generated ID into the HTML which frontend can include in subsequent requests). Session cookie is a nice idea but it's too small. Backend global cache (e.g. memorystore) is nice but expensive. Backend database is too "heavy" and slow. The right solution feels like using browser local storage / session storage (even better because this has "tab id" stuff built-in). Indeed, this is how I understand most frontend frameworks work. React makes use of session storage.

My question is how I can interact with session storage. I think backend can only return information to client and client must do the interaction. I think this must be by javascript or hyperscript. Am I correct?

**Reply by Telroshan** [reply]
Ooooh ok I understand now, thanks for the details!
Yeah htmx doesn't provide any session storage utility as it's a pure client-side thing and htmx is really meant for server-side communication.
I'm not a hyperscript user so I couldn't tell you if it already exists in the lib or not, but that's the kind of thing I would expect it to be able to do. The [cheatsheet](https://hyperscript.org/img/hyperscript-cheatsheet.pdf) references the following
> localStorage.foo value in local storage

sessionStorage isn't referenced in that doc, but I would expect it to be available as well
Otherwise indeed, vanilla JS will do the trick!

In that specific use case you described though, I think I would have personnally used hx-include on the table to include any "staging form" or inputs that are inside. Then get the values from the backend, and restore the "in progress state" from there.
It's rough but here's the idea I was thinking about in [this JSFiddle](https://jsfiddle.net/c0xerhq9/)
- `hx-include` on the headers to include `closest table`, which will include any input value defined in the table, in the request _(imagining clicking on the headers would sort the rows here)_
- `hx-get` on the initial rows, that bring in an editable version when clicking on them _(values are not retrieved here ofc since you would get them from the DB typically, but you get the idea)_
- In the editable rows, you define a hidden input for the ID of that row, and make each column an input
- Now, when you click on the headers, all "in-progress" rows will be sent along, **in order**, to the backend
![image](https://github.com/bigskysoftware/htmx/assets/19146183/710c143c-575a-4385-a3c4-ba3b3c6bb43d)
- From there, I would retrieve those form values in my backend, and set a boolean in the data I pass _(to render the HTML)_ to render either the initial row or the editable one. Also override the row's values with the "staging" ones before rendering
- The benefit here is that you don't even need to bother using session storage or anything, everything is directly in the HTML, and automatically sent to the backend. But I guess it's a matter of preference here, I prefer moving that kind of work on the server, but you may totally stick to client side logic and use session storage instead!

Hope that makes sense, and helps 😆 

[Edit : updated the JSFiddle URL as I of course forgot to save the latest changes]

**Reply by adamcunnington-mlg** [reply]
Amazing! I will fully digest this but this is much appreciated - kind of you, thank you.
